### DISCUSSION TITLE
Why doesn't console.status(...) render correctly with tons of text from a subprocess

### DISCUSSION THREAD

**Discussion by Hubro** [Textualize/rich#2886] [Q&A] [answered]
```python3
with rich_console.status("Building Python inside Docker container"):
    subprocess.check_call(
        [
            "docker",
            "run",
            "--rm",
            "-u",
            f"{os.getuid()}:{os.getgid()}",
            "-v",
            f"{tmpdir}:/workdir",
            "--entrypoint",
            "bash",
            f"python-build:{environment}",
            "-c",
            BUILD_SCRIPT.format(python_version=python_version),
        ]
    )
```

Basically I'm building Python inside a Docker container in a subprocess. `subprocess.check_call` pipes all output from the child process to `sys.stdout` and `sys.stderr`, as far as I'm aware. The result:

![image](https://user-images.githubusercontent.com/597206/226489075-e64d0c75-1988-48cb-b632-8f06f1ebb303.png)

I would expect the status text and spinner to always be at the bottom of the screen. Am I using it wrong? Or does it just not work in scenarios where tons of text is written to the terminal really fast?

**Comment by willmcgugan** [comment, marked as answer]
The child process is writing directly to stdout, which means that the main process is unable to intercept the output.

If you want a subprocess to work with Rich status, you will need to capture and print the output from the child process.

**Reply by Hubro** [reply]
That makes sense. Does Rich provide a file-like interface anywhere for writing into the console? That seems extremely useful. That way I can just pass that interface straight into the subprocess call, something like:

```python3
subprocess.check_call(
    [...],
    stdout=rich_console.file_wrapper,
    stderr=rich_console.file_wrapper,
)
```

**Reply by Hubro** [reply]
I ended up just writing a wrapper that does as you say, and it works nicely:

```python3
def run_subprocess(status_msg: rich.console.RenderableType, command: list[str]) -> None:
    """Run a subprocess while displaying a status message using Rich"""

    with console.status(status_msg):
        process = subprocess.Popen(
            command,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            encoding="UTF-8",
        )

        assert process.stdout

        for line in process.stdout:
            console.out(line, end="", highlight=False)
```

**Reply by willmcgugan** [reply]
Rich will capture stdout, so you should just be able to set stdout and stderr to `sys.stdout` and `sys.stderr`

**Reply by Hubro** [reply]
I tried that yesterday, but it resulted in the original behavior. I'm guessing Python is "smart" enough to actually forward the file handles to the subprocess and not capture the output.
